### ISSUE TITLE
Do not supply -Wbitwise-op-parentheses by default

### ISSUE THREAD

**Issue by d6fd0808-f213-4703-8fea-222e641ecd67** [llvm/llvm-project#27818]
|  |  |
| --- | --- |
| Bugzilla Link | [27444](https://llvm.org/bz27444) |
| Version | 3.7 |
| OS | FreeBSD |
| CC | @zygoloid |

## Extended Description 
clang seems to have -Wbitwise-op-parentheses enabled by default, causing hundreds of suprious warnings to appear when compiling perfectly fine code. I request that clang should not warn about things like "&& within ||" or "& within |" by default as these things appear often in normal C code and are usually not an indication of incorrect code. After all, most C programmers are able to remember that & has higher precedence than | by analogy to * and + as has been the case in mathematics since the 18th century.

**Comment by Quuxplusone** [comment]
Presumably the "false positive" here is
```
bool f(int a, int b, int c)
{
    return (a==1) || (b==2) && (c==3);
}
```
Both Clang and GCC warn with `-Wall`; neither Clang nor GCC warn without `-Wall`. (And this code is definitely worth warning about, IMO!)

Since Clang doesn't warn without `-Wall`, I think this is either not-a-bug or fixed-sometime-between-2016-and-2021.

**Comment by clausecker** [comment]
@Quuxplusone This issue is still current as clang still warns for the same situation involving `&` and `|`.  See my example for bug #30763 for an example that triggers this.  Again it is surprising to me that you just close issues without asking for details if you don't understand what the problem is.  That is not nice.

**Comment by dwblaikie** [comment]
Doesn't look to me like this warning is enabled by default. Could you provide a specific example? (source code, clang version, clang command line, and diagnostic output)
eg: https://godbolt.org/z/snWMG8cfo https://godbolt.org/z/sfEYaj9rq

And the Clang diagnostic documentation: https://clang.llvm.org/docs/DiagnosticsReference.html#wlogical-op-parentheses (diagnostics that might be enabled by default include the text "Some of the diagnostics controlled by this flag are enabled by default." but that text is not present for the "bitwise-op-parentheses" group). And the Clang source code https://github.com/llvm/llvm-project/blob/6f61fe7de97c78adc8fac528239260122fb83be5/clang/include/clang/Basic/DiagnosticSemaKinds.td#L6503

And yeah, generally with a somewhat vague (& also somewhat snarky re: "most C programmers ... " etc) bug, closing it summarily with a rough guess at what was meant seems fine to me in this community - it can be readily reopened if more data helps clarify things.
